_source/_posts/2023-07-27-spring-graphql-react.md

```diff
@@ -0,0 +1,1794 @@
+✔ Would you like to use Tailwind CSS? ... No
+✔ Would you like to use `src/` directory? ... Yes
+✔ Would you like to use App Router? (recommended) ... Yes
+✔ Would you like to customize the default import alias? ... No
+```
+
+Then add the the MUI Datagrid dependencies and custom hooks from Vercel:
+
+```shell
+cd react-graphql && \
+  npm install @mui/x-data-grid && \
+  npm install @mui/material@5.14.5 @emotion/react @emotion/styled && \
+  npm install react-use-custom-hooks
+```
+
+Test run the application with:
+
+```shell
+npm run dev
+```
+
+Navigate to http://localhost:3000 and you should see the default Next.js page:
+
+{% img blog/spring-graphql-react/nextjs-default.png alt:"Next.js default page" width:"800" %}{: .center-image }
+
+### Create the API client
+
+Create the folder `src/services` and add the file `base.tsx` with the following code:
+
+{% raw %}
+__base.tsx__
+```tsx
+import axios from "axios";
+
+export const backendAPI = axios.create({
+  baseURL: process.env.NEXT_PUBLIC_API_SERVER_URL
+});
+
+export default backendAPI;
+```
+{% endraw %}
+
+Add the file `src/services/companies.tsx` with the following content:
+
+{% raw %}
+__companies.tsx__
+```tsx
+import { AxiosError } from "axios";
+import { backendAPI } from "./base";
+
+export type CompaniesQuery = {
+  page: number;
+};
+
+export type CompanyDTO = {
+  name: string;
+  SIC: string;
+  id: string;
+  companyNumber: string;
+  category: string;
+};
+
+export const CompanyApi = {
+
+  getCompanyCount: async () => {
+    try {
+      const response = await backendAPI.post("/graphql", {
+        query: `{
+        companyCount
+      }`,
+      });
+      return response.data.data.companyCount as number;
+    } catch (error) {
+      console.log("handle get company count error", error);
+      if (error instanceof AxiosError) {
+        let axiosError = error as AxiosError;
+        if (axiosError.response?.data) {
+          throw new Error(axiosError.response?.data as string);
+        }
+      }
+      throw new Error("Unknown error, please contact the administrator");
+    }
+  },
+
+  getCompanyList: async (params?: CompaniesQuery) => {
+    try {
+      const response = await backendAPI.post("/graphql", {
+        query: `{
+        companyList(page: ${params?.page || 0}) {
+          name,
+          SIC,
+          id,
+          companyNumber,
+          category
+        }}`,
+      });
+      return response.data.data.companyList as CompanyDTO[];
+    } catch (error) {
+      console.log("handle get companies error", error);
+      if (error instanceof AxiosError) {
+        let axiosError = error as AxiosError;
+        if (axiosError.response?.data) {
+          throw new Error(axiosError.response?.data as string);
+        }
+      }
+      throw new Error("Unknown error, please contact the administrator");
+    }
+  },
+
+};
+```
+{% endraw %}
+
+Add a file `.env.example` and `.env.local` in the root folder, both with the following content:
```

Comment:
Why are both necessary? And should one be added to `.gitignore`?

Reply:
Added a note about the files.